blockchain/chain.go

```diff
@@ -1143,94 +1167,148 @@ func (b *BlockChain) reorganizeChainInternal(targetTip *blockNode) error {
-// without flushing in the case the chain is not able to reorganize due to a
-// block failing to connect.
+// without flushing.
 //
 // This function MUST be called with the chain state lock held (for writes).
-func (b *BlockChain) reorganizeChain(targetTip *blockNode) error {
-	// Nothing to do if there is no target tip or the target tip is already the
-	// current tip.
-	if targetTip == nil {
-		return nil
-	}
-	origTip := b.bestChain.Tip()
-	if origTip == targetTip {
+func (b *BlockChain) reorganizeChain(target *blockNode) error {
+	// Nothing to do if there is no target specified or it is already the
+	// current best chain tip.
+	tip := b.bestChain.Tip()
+	if target == nil || tip == target {
 		return nil
 	}
+	origTip := tip
 
-	// Send a notification announcing the start of the chain reorganization.
-	b.chainLock.Unlock()
-	b.sendNotification(NTChainReorgStarted, nil)
-	b.chainLock.Lock()
+	var sentReorgingNtfn bool
+	var reorgErrs []error
+	for ; target != nil && tip != target; tip = b.bestChain.Tip() {
+		select {
+		case <-b.interrupt:
+			return errInterruptRequested
+		default:
+		}
 
-	defer func() {
-		// Send a notification announcing the end of the chain reorganization.
-		b.chainLock.Unlock()
-		b.sendNotification(NTChainReorgDone, nil)
-		b.chainLock.Lock()
-	}()
-
-	// Attempt to reorganize to the chain to the new tip.  In the case it fails,
-	// reorganize back to the original tip.  There is no way to recover if the
-	// chain fails to reorganize back to the original tip since something is
-	// very wrong if a chain tip that was already known to be valid fails to
-	// reconnect.
-	//
-	// NOTE: The failure handling makes an assumption that a block in the path
-	// between the fork point and original tip are not somehow invalidated in
-	// between the point a reorged chain fails to connect and the reorg back to
-	// the original tip.  That is a safe assumption with the current code due to
-	// all modifications which mark blocks invalid being performed under the
-	// chain lock, however, this will need to be reworked if that assumption is
-	// violated.
-	fork := b.bestChain.FindFork(targetTip)
-	reorgErr := b.reorganizeChainInternal(targetTip)
-	if reorgErr != nil {
-		if err := b.reorganizeChainInternal(origTip); err != nil {
-			panicf("failed to reorganize back to known good chain tip %s "+
-				"(height %d): %v -- probable database corruption", origTip.hash,
-				origTip.height, err)
+		// Determine if the chain is being reorganized to a competing branch.
+		// This is the case when the current tip is not an ancestor of the
+		// target tip.
+		isReorganizing := target.Ancestor(tip.height) != tip
+		if isReorganizing && !sentReorgingNtfn {
+			// Send a notification announcing the start of the chain
+			// reorganization.
+			//
+			// Notice that the chain lock is not released before sending the
+			// notification.  This is intentional and must not be changed
+			// without understanding why!
```

Comment:
The chain lock is held.  Notice the `reorganizeChain` function requires it.  The comment is repeated because it applies to each notification.  Some of them require it to be released and some of them don't.  So, I wanted to be super explicit about which ones do not release it.

Yes, I didn't explain why so that people won't touch it under they understand why things are that way!